src/main/java/uk/gov/ons/census/action/schedule/ActionInstructionBuilder.java

```diff
@@ -0,0 +1,119 @@
+    } else if (uacQidLinks.size() > 1) {
+      if (isQuestionnaireWelsh(caze.getTreatmentCode()) && uacQidLinks.size() == 2) {
+        uacQidLink = getSpecificUacQidLinkByQuestionnaireType(uacQidLinks, "02", "03");
+        uacQidLinkWales =
+            Optional.ofNullable(getSpecificUacQidLinkByQuestionnaireType(uacQidLinks, "03", "02"));
+      } else {
+        throw new RuntimeException(); // TODO: How do we know which one to use?
+      }
+    } else if (!isQuestionnaireWelsh(caze.getTreatmentCode())) {
+      uacQidLink = uacQidLinks.get(0);
+    } else {
+      // Not enough UAC/QID links for a Welsh questionnaire
+      throw new RuntimeException();
+    }
+
+    ActionEvent actionEvent = new ActionEvent();
+    actionEvent
+        .getEvents()
+        .add("CASE_CREATED : null : SYSTEM : Case created when Initial creation of case");
+
+    ActionAddress actionAddress = new ActionAddress();
+    actionAddress.setLine1(caze.getAddressLine1());
+    actionAddress.setLine2(caze.getAddressLine2());
+    actionAddress.setLine3(caze.getAddressLine3());
+    actionAddress.setTownName(caze.getTownName());
+    actionAddress.setPostcode(caze.getPostcode());
+    actionAddress.setOrganisationName(caze.getOrganisationName());
+    ActionRequest actionRequest = new ActionRequest();
+    actionRequest.setActionId(UUID.randomUUID().toString());
+    actionRequest.setResponseRequired(false);
+    actionRequest.setActionPlan(actionRule.getActionPlan().getId().toString());
+
+    actionRequest.setActionType(actionRule.getActionType().toString());
+
+    actionRequest.setAddress(actionAddress);
+    actionRequest.setLegalBasis("Statistics of Trade Act 1947");
+    actionRequest.setCaseGroupStatus("NOTSTARTED");
+    actionRequest.setCaseId(caze.getCaseId().toString());
+
+    actionRequest.setPriority(Priority.MEDIUM);
+    actionRequest.setCaseRef(Long.toString(caze.getCaseRef()));
+    actionRequest.setIac(uacQidLink.getUac());
+    actionRequest.setQid(uacQidLink.getQid());
+
+    if (uacQidLinkWales.isPresent()) {
+      actionRequest.setIacWales(uacQidLinkWales.get().getUac());
+      actionRequest.setQidWales(uacQidLinkWales.get().getQid());
+    }
+
+    actionRequest.setEvents(actionEvent);
+    actionRequest.setExerciseRef("201904");
+    actionRequest.setUserDescription("Census-FNSM580JQE3M4");
+    actionRequest.setSurveyName("Census-FNSM580JQE3M4");
+    actionRequest.setSurveyRef("Census-FNSM580JQE3M4");
+    actionRequest.setReturnByDate("27/04");
+    actionRequest.setSampleUnitRef("DDR190314000000516472");
+    ActionInstruction actionInstruction = new ActionInstruction();
+    actionInstruction.setActionRequest(actionRequest);
+
+    return actionInstruction;
+  }
+
+  private boolean isQuestionnaireWelsh(String treatmentCode) {
+    return (treatmentCode.startsWith("HH_Q") && treatmentCode.endsWith("W"));
+  }
+
+  private UacQidLink getSpecificUacQidLinkByQuestionnaireType(
```

Comment:
This logic seems mixed with the previous parts of uacQidLinks.   
I'm sure the whole uacQuidLink could all be written in a more readable.